### PR TITLE
refactor(swing-store): Use the same stream-based approach for loading snapshots as saving them

### PR DESCRIPTION
**Pull request by gibson042**
## Description

Copies the `const cleanup = []; return aggregateTryFinally(…)` pattern and removes consequently-unused helpers.

We could try abstracting that pattern itself into a helper, but I don't think it would save much because the particulars are different.

### Security Considerations

n/a

### Documentation Considerations

n/a

### Testing Considerations

n/a

### CODE REVIEW

File: packages/swing-store/src/snapStore.js [hunk 1]
```diff
@@ -344,13 +269,46 @@ export function makeSnapStore(
...
+          cleanup: tmpCleanup,
+        } = await ptmpFile({ template: `load-${hash}-XXXXXX.xss` });
+        cleanup.push(tmpCleanup);
+        const snapWriter = createWriteStream(noPath, {
+          fd,
+          autoClose: false,
+        });
+        cleanup.push(() => snapWriter.close());
+
+        await Promise.all([gzReader, snapWriter].map(s => fsStreamReady(s)));
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js]: Nit: I'm not convinced by the readability of `.map` in this case

File: packages/swing-store/src/snapStore.js [hunk 2]
```diff
@@ -344,13 +269,46 @@ export function makeSnapStore(
...
+        const hashStream = createHash('sha256');
+        snapReader.pipe(hashStream);
+        snapReader.pipe(snapWriter);
+
+        await Promise.all([gzReader, snapWriter].map(s => finished(s)));
+        const h = hashStream.digest('hex');
+        h === hash || assert.fail(d`actual hash ${h} !== expected ${hash}`);
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js:299]: I did not realize we were checking the hash of the decompressed file!

File: packages/swing-store/src/snapStore.js [hunk 3]
```diff
@@ -344,13 +269,46 @@ export function makeSnapStore(
...
+        await pfsync(fd);
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js]: I believe that doing an `fsync` in this case is not necessary, and probably counter productive. We don't actually care if things have been flushed to disk, and would even prefer if it stayed in memory. The OS should guarantee that a different process reading this file should get the content we just wrote without it having been flushed to the underlying device.

File: packages/swing-store/src/snapStore.js [hunk 4]
```diff
@@ -99,7 +97,7 @@ export function makeSnapStore(
     createWriteStream,
     measureSeconds,
     fsync,
-    open,
+    open: _open,
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js]: Can't we simply remove this?

**Review by mhofman — COMMENTED** [review]
Nice cleanup, but I think the `fsync` should be removed.

**Comment by gibson042** [comment]
@mhofman Back to you.

File: packages/swing-store/src/snapStore.js [hunk 5]
```diff
@@ -318,26 +320,26 @@ export function makeSnapStore(
...
+            : undefined);
         }
-      }),
-    );
+      : async hash => {
+          // Update toDelete iff there is no error from unlink.
+          const fullPath = fullPathFromHash(hash);
+          await (keepSnapshots !== true ? unlink(fullPath) : undefined);
+          toDelete.delete(hash);
+        };
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js]: Edit: yikes GitHub is messing up the diff pretty bad!

```suggestion
    const handleError = ignoreErrors ? p => p.catch(sink) : p => p;
    const deleteHash = async hash => {
        const fullPath = fullPathFromHash(hash);
        await (keepSnapshots !== true ? handleError(unlink(fullPath)) : undefined);
        // Update toDelete iff there is a non-ignored error from unlink.
        toDelete.delete(hash);
      };
```

**Review by mhofman — COMMENTED** [review]
Nit to DRY it up

File: packages/swing-store/src/snapStore.js (hunk 5 above)
Comment (gibson042) [reply packages/swing-store/src/snapStore.js]: Nice; done.

**Review by mhofman — APPROVED** [review]

File: packages/swing-store/src/snapStore.js [hunk 6]
```diff
@@ -331,9 +331,9 @@ export function makeSnapStore(
     };
 
     const results = await Promise.allSettled([...toDelete].map(deleteHash));
-    const errors = results.flatMap(({ status, reason }) =>
-      status === 'rejected' ? [reason] : [],
-    );
+    const errors = /** @type {PromiseRejectedResult[]} */ (
+      results.filter(({ status }) => status === 'rejected')
+    ).map(({ reason }) => reason);
```
Comment (mhofman) [inline packages/swing-store/src/snapStore.js:336]: Hoping that future versions of TS will help with cases like this.

**Pull request merged**